### PR TITLE
Big arrays sliced from netty buffers (byte)

### PR DESCRIPTION
**Pull request by martijnvg**
Based on #90745 but for longs.
This should allow aggregations down the road to read long values directly from netty buffer, rather than copying it from the netty buffer.

Relates to #89437

### CODE REVIEW

**Review by nik9000 — APPROVED** [review]
Seems right to me.

**Review by not-napoleon — APPROVED** [review]
This looks right to me. As with all of these, I think we should let the distributed team know we're adding it in.

File: server/src/main/java/org/elasticsearch/common/util/ReleasableByteArray.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    }
+
+    @Override
+    public boolean hasArray() {
+        return ref.hasArray();
+    }
+
+    @Override
+    public byte[] array() {
+        return ref.array();
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/common/util/ReleasableByteArray.java:84]: This seems dangerous. There's a weird assumption in the arrays returned by this method in so far that we assume that valid data starts at index `0`.
This is not guaranteed for arrays resulting from `in.readReleasableBytesReference();` (as a matter of fact, it's rather unlikely.

If we want to support this, which seems nice, we'd have to enhance the interface to return an `arrayOffset` just like we do with `org.elasticsearch.common.bytes.BytesReference#arrayOffset`.

**Review by original-brownbear — CHANGES_REQUESTED** [review]
This looks just fine except for the rather serious issue with the `array()` method that we have to resolve by either not supporting it or enhancing the interface IMO.

File: server/src/main/java/org/elasticsearch/common/util/ReleasableByteArray.java (hunk 1 above)
Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/common/util/ReleasableByteArray.java:84]: Thanks for spotting this! Maybe for now shouldn't support this? I can add a comment here and throw unsupported operation exception. Then if/when this really gets used the interface can be improved and supported in a safe manner.

Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/common/util/ReleasableByteArray.java:84]: Sounds good, lets make it not supported here for now then :)
This was used as an optimization for the path `ByteArray` -> `BytesReference` which is quite nice. I can open a suggested fix PR once this is in :)

File: server/src/main/java/org/elasticsearch/common/util/ReleasableByteArray.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    }
+
+    @Override
+    public boolean hasArray() {
+        return ref.hasArray();
+    }
+
+    @Override
+    public byte[] array() {
+        // The assumption of this method is that the returned array has valid entries starting from sloy 0 and
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/common/util/ReleasableByteArray.java:84]: I guess you mean, "slot" here? :)

**Review by original-brownbear — APPROVED** [review]

**Comment by martijnvg** [comment]
@elasticsearchmachine update branch

**Comment by martijnvg** [comment]
@elasticmachine update branch

**Pull request merged**